### PR TITLE
Write thumbnails in one pass with atomic rename

### LINKED ISSUE #10: Move to destination not atomic
When creating thumbnails, thumbnailer creates temporary files in what is returned by `std::env::temp_dir()` (this will generally be /tmp), and then runs the `mv` command to move them into place.

Move can only be atomic within the same filesystem, using the rename system call. As /tmp is commonly tmpfs, the move will commonly be crossing filesystem boundaries, this is not possible, and the `mv` command ends up doing a copy instead.

This renders the entire operation moot, as this copy is no better than just writing straight to the destination in the first place.

The way this is usually dealt with is to use the destination for the temp file using temporary name (e.g. `<MD5_OF_URI>_tmp.png` or `<MD5_OF_URI>.png.tmp`).

### PR DESCRIPTION
**Pull request by kennylevinsen**
Previously, the thumbnail was written to /tmp, read back, re-written to
/tmp and finally copied to the destination folder. This lead to many
problems, especially with certain FUSE filesystems.

Instead, write the thumbnail in a single pass to a temporary file in the
destination folder, and rename it when done to the destination file.
This avoids wasteful I/O and increases the chances of the rename being
atomic.

Closes: https://github.com/filcuc/thumbnailer/issues/10

Note: Contains https://github.com/filcuc/thumbnailer/pull/11.

### CODE REVIEW

**Comment by kennylevinsen** [comment]
Rebased :)

File: src/thumbnailer.rs [hunk 1]
```diff
@@ -195,87 +201,91 @@ impl Thumbnailer {
...
+        let temp_path = format!("{}.tmp", thumbnailer.destination_path.to_str().unwrap());
+        let output = std::fs::OpenOptions::new()
+            .write(true)
+            .truncate(true)
+            .create(true)
+            .open(&temp_path)
+            .map_err(|e| format!("Failed to open thumbnailer in temporary dir: {}", e))?;
+
+        let thumbnail = thumbnailer.thumbnail.as_ref().unwrap();
+        let (ct, bits) = match thumbnail.color() {
```
Comment (filcuc) [inline src/thumbnailer.rs:213]: Do we really need to get low level and query the original image format?

File: src/thumbnailer.rs [hunk 2]
```diff
@@ -195,87 +201,91 @@ impl Thumbnailer {
...
+            .map_err(|e| format!("Error writing PNG header: {}", e))?;
 
         let uri_raw = Thumbnailer::calculate_path_uri(
             thumbnailer.use_full_path_for_md5,
             &thumbnailer.source_path,
         );
-        let uri = crate::png::Chunk::new_text("Thumb::URI", uri_raw).unwrap();
-        chunks.insert(1, uri);
+        writer
+            .write_chunk(PNG_TEXT_KIND, &text_chunk("Thumb::URI", uri_raw).unwrap())
```
Comment (filcuc) [inline src/thumbnailer.rs:236]: Curious: were all these chunk related functions available in previous versions of the png library? Is this the reason for the upgrade?

File: src/thumbnailer.rs (hunk 1 above)
Comment (kennylevinsen) [reply src/thumbnailer.rs:213]: Yeah, the image-to-png type mapping is a little ugly.

To write the png in one pass, we need to get a hold of the PNG encoder to write the metadata chunks. This rules out image's save_with_format, as it doesn't expose the encoder. Instead, I just [took this image-to-png type mapping straight from the image lib](https://docs.rs/image/0.23.6/src/image/png.rs.html#508).

File: src/thumbnailer.rs (hunk 2 above)
Comment (kennylevinsen) [reply src/thumbnailer.rs:236]: Tbh, the dep updates were just done as part of the png crate research. I had initially removed the png dependency (it wasn't used at all), and then when I ended up needing it I added the latest version.

Want me to remove or isolate the version bumps? I'm fine either way.

**Pull request merged**